### ISSUE TITLE
Support type parameterized `Deserialize` function 

### ISSUE THREAD

**Issue by 1ambda** [weichx/cerialize#9]
I am writing `Dserializable` abstract class which indicate it can be **deserialized** and has `deserialize` Function. It looks like

```typescript
const cerialize = require('cerialize'); // import

abstract class Deserializable {
  public static deserialize<T> (json, constructor): T {
    return cerialize.Deserialize(json, constructor);
  }
}

// usage
class Command extends Deserializable {
  @deserialize public help: boolean;
  @deserializeAs(CommandOrder, "_order") public orders: Array<CommandOrder>;
  @deserializeAs("_args") public args: Array<String>;
}

let cmd = Command.deserialize<Command>(json, Command);
```

But since [`cerialize.Deserialize`](https://github.com/weichx/cerialize/blob/87fe32f40e29abb97055e8f6b79be13fc73c597c/src/serialize.ts#L404) requires the constructor function as the second arg, It's somewhat verbose. 

If we have a type parameterized `cerialize.Deserialize` function, we can make more simplified  `Deserialzable`. For example,

```typescript
export abstract class Deserializable {
  public static deserialize<T> (json): T {
    return Deserialize<T>(json);
  }
}

// more simple
Command.deserialize<Command>(json);
```

Thanks. :)

**Comment by weichx** [comment]
How would you get the run time type information to know what to type to deserialize something into? I'm all for using generics here but since the generic constraints are a compile time feature of Typescript and not reflected in the compiled JS I'm not sure you can make this work in the same way you'd solve this problem in Java or C#.

**Comment by weichx** [comment]
Maybe if you use this it would work but I'd have to try it
```Typescript
export abstract class Deserializable {
  public static deserialize<T> (json): T {
    return Deserialize<T>(json, this.constructor);
  }
}
```

I'm also not sure that static methods are inherited, so it might need to be more like:
```Typescript
export abstract class Deserializable {
  public deserialize (json : any): this {
    return DeserializeInto(json, this.constructor, this);
  }
}
var cmd = new Command().deserialize(json);
```
Thats using polymorphic `this` which landed here: https://github.com/Microsoft/TypeScript/pull/4910 but I haven't tried to use this feature before so im not sure if it works as I expect.
This kind of thing really wants to be a mixin, can't wait for that to land in typescript.

**Comment by weichx** [comment]
@1ambda any thoughts here? I'm going to close this otherwise

**Comment by 1ambda** [comment]
@weichx Thanks for detailed explanation and alternatives. 

1. public static methods are not inherited in TypeScript 
2. polymorphic `this` also requires type parameter for example, 

```typescript
export abstract class Deserializable<T> {
  public deserialize<T>(json):T {
    return cerialize.Deserialize(json, this);
  }
}

class Car extends Deserializable2<Car> {
  @deserialize public engine: string;
  @deserialize public wheels: number;
}

let json = {engine: "M5", wheels: 4};
let c1: Car = new Car().deserialize<Car>(json); // compile error
let c1: Car = new Car().deserialize(json); // compiles
```

I read https://github.com/Microsoft/TypeScript/issues/2037 but could not found any solution. 
I agree this is the limitation of TypeScript as you mentioned.

My final implementation looks like 

```typescript
export interface NoParamConstructor<T> {
  new (): T
}

export abstract class Deserializable {
  public static deserialize<T>(ctor: NoParamConstructor<T>, json): T {
    return cerialize.Deserialize(json, ctor);
  }

  public static deserializeArray<T>(ctor: NoParamConstructor<T>, json): Array<T> {
    return cerialize.Deserialize(json, ctor);
  }
}
```

Without NoParamConstructor, we have to pass type parameter explicitly. for example, 

```typescript
let json = {engine: "M5", wheels: 4};
let c1 = Car.deserialize(Car, json);              // with NoParamConstructor
let c2 = Car.deserialize<Car>(Car, json);    // without NoParamConstructor
```

I will close this issue Thanks again :)

**Comment by weichx** [comment]
@1ambda how were you using `deserialize`? I added this PR https://github.com/weichx/cerialize/pull/13 with your test in it and using `deserialize` and it passes just fine for me

**Comment by weichx** [comment]
Another thought on the above code, it seems like `Car.deserialize<Car>(Car, json)` is a bit odd. I played with generics a bit more and came up with this:
```typescript
function DeserializeGeneric<T>(json : any, type : T) : T {
	return <T>Deserialize(json, type);
}
```
which you should be able to use like this:
```typescript
var car = DeserializeGeneric(json, Car);
```

This is probably something I should add to the library, but for now I think its a cleaner solution for your use case than extending an abstract class.

**Comment by 1ambda** [comment]
@weichx I downloaded `0.1.2` and checked that the test is passing as you pointed out. I think the reason why `deserialize` does not work properly only in my code is, I used customized version of **cerialize**  in `node_modules/cerialize` which i modified since the previous released version (`0.1.1`) does not support #11.

In terms of `DeserializeGeneric<T>`, I agree with your suggestion. I will use it thanks :+1:  

Lastly, It would be good if you modify `Deserialize` signature to use generic type `T`. Current implementations looks like

```typescript
export function Deserialize(json : any, type? : Function|ISerializable) : any {
```

As you mentioned, it could be 

```typescript
// If we use NoParamConstructor technique introduced Microsoft/TypeScript#2037 it would be more safe.
// For example, `type: NoParamConstructor<T>` instead of just specifying `T`  

export function Deserialize<T>(json : any, type: T) : T {
```

Thanks :)
